platooning_tactical_plugin/src/platooning_tactical_plugin/platooning_tactical_plugin.cpp

```diff
@@ -0,0 +1,192 @@
+
+        ROS_INFO("Successfully launched node.");
+    }
+
+    PlatooningTacticalPlugin::~PlatooningTacticalPlugin()
+    {
+    }
+
+    bool PlatooningTacticalPlugin::readParameters()
+    {
+        nodeHandle_.param("trajectory_time_length", trajectory_time_length_, 6.0);
+        nodeHandle_.param("trajectory_point_spacing", trajectory_point_spacing_, 0.1);
+        return true;
+    }
+
+    bool PlatooningTacticalPlugin::plan_trajectory_cb(cav_srvs::PlanTrajectoryRequest &req, cav_srvs::PlanTrajectoryResponse &resp){
+
+        if(req.maneuver_plan.maneuvers[0].lane_following_maneuver.parameters.neogition_type == 2 && req.maneuver_plan.maneuvers[0].lane_following_maneuver.parameters.planning_strategic_plugin == "PlatooningStrategicPlugin" ){
+            if(req.maneuver_plan.maneuvers[0].lane_following_maneuver.start_speed == req.maneuver_plan.maneuvers[0].lane_following_maneuver.end_speed ) {
+                resp.trajectory_plan = trajectory_msg;
+                resp.related_maneuvers.push_back(cav_msgs::Maneuver::LANE_FOLLOWING);
+                resp.maneuver_status.push_back(cav_srvs::PlanTrajectory::Response::MANEUVER_IN_PROGRESS);
+            }
+        }
+        else {
+            resp.trajectory_plan = trajectory_msg;
+            resp.related_maneuvers.push_back(cav_msgs::Maneuver::LANE_FOLLOWING);
+            resp.maneuver_status.push_back(cav_srvs::PlanTrajectory::Response::MANEUVER_IN_PROGRESS);
+        }
+
+        return true;
+    }
+
+    void PlatooningTacticalPlugin::waypoints_cb(const autoware_msgs::LaneConstPtr& msg)
+    {
+        if(msg->waypoints.size() == 0)
+        {
+            ROS_WARN_STREAM("Received an empty waypoints!");
+            return;
+        }
+
+        cav_msgs::TrajectoryPlan trajectory;
+        trajectory.header.frame_id = msg->header.frame_id;
+        trajectory.header.stamp = ros::Time::now();
+        trajectory.trajectory_id = boost::uuids::to_string(boost::uuids::random_generator()());
+        trajectory.trajectory_points = compose_trajectory_from_waypoints(msg->waypoints);
+        waypoints_list = msg->waypoints;
+        trajectory_msg = trajectory;
+
+        
+        ROS_DEBUG("I heard waypoints");
+    };
+
+
+    std::vector<cav_msgs::TrajectoryPlanPoint> PlatooningTacticalPlugin::compose_trajectory_from_waypoints(std::vector<autoware_msgs::Waypoint> waypoints)
+    {
+        std::vector<autoware_msgs::Waypoint> partial_waypoints = get_waypoints_in_time_boundary(waypoints, trajectory_time_length_);
+        std::vector<cav_msgs::TrajectoryPlanPoint> tmp_trajectory = create_uneven_trajectory_from_waypoints(partial_waypoints);
+        std::vector<cav_msgs::TrajectoryPlanPoint> final_trajectory = post_process_traj_points(tmp_trajectory);
+        return final_trajectory;
+    }
+
+    std::vector<cav_msgs::TrajectoryPlanPoint> PlatooningTacticalPlugin::create_uneven_trajectory_from_waypoints(std::vector<autoware_msgs::Waypoint> waypoints)
```

Comment:
In general, use the speed received from the trajectory plan request instead of waypoint speed. You can create a variable double traj_speed = (start_speed+end_speed)/2 and use min(watpoint_speed, traj_speed) to generate a trajectory.